File: Week_01/42/trapping_rain_water.py
from typing import List


class Solution:
    def trap(self, height: List[int]) -> int:

        stack = []
        res = 0

        for i, h in enumerate(height):
            while stack and height[stack[-1]] < h:
                top = stack.pop()
                if not stack:
                    break
                left = stack[-1]
                res += (i - left - 1) * (min(h, height[left]) - height[top])
            stack.append(i)
        return res

    def test(self):
        for height, target in [
            ([], 0),
            ([1, 1], 0),
            ([1, 0, 0], 0),
            ([1, 0, 1], 1),
            ([0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1], 6),
        ]:
            ans = self.trap(height)
            assert ans == target, f"target: {target}, ans {ans}"
        print("all done")


if __name__ == "__main__":
    Solution().test()

File: Week_01/42/trapping_rain_water.go
package main

func trap(height []int) int {
	stack := make([]int, 0)
	res := 0

	for i, h := range height {
		for len(stack) > 0 && height[stack[len(stack)-1]] < h {
			top := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			if len(stack) == 0 {
				break
			}
			left := stack[len(stack)-1]
			res += (i - left - 1) * (min(h, height[left]) - height[top])
		}
		stack = append(stack, i)
	}
	return res
}

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}
